src/find-options/FindConditions.ts

```diff
@@ -3,6 +3,9 @@ import {FindOperator} from "./FindOperator";
 /**
  * Used for find operations.
  */
-export type FindConditions<T> = {
-    [P in keyof T]?: T[P] extends never ? FindConditions<T[P]>|FindOperator<FindConditions<T[P]>> : FindConditions<T[P]>|FindOperator<FindConditions<T[P]>>;
+type FindCondition<T> = FindConditions<T> | FindOperator<FindConditions<T>>;
+export type FindConditions<T> = T | {
```

Comment:
I don't think I remember why we had `T[P] extends never`, but since we have it most probably we had some issue. But why did you add `T | ` ? Isn't second part already cover T?

Reply:
`T[P] extends never` is to workaround the typescript bug: https://github.com/typeorm/typeorm/pull/4470.

`T | ` is required because of weird behavior of `{[P keyof T]}` for generics. Example:

```
type A<T> = {[P in keyof T]: T[P]}

type B = A<string>;
```

The computed type of `B` is `string` which is counter intuitive.
So such magic doesn't work for non-generic type which will always result in recursion types call in previous implementation. So I can not write the condition like: `{category: category}` becuase `category` is not generic. The thing I can only do without `T |` is the following `{category: {id: category.id}}` which is working btw. Let me know if you want such query to also be added to the test if it is not tested yet in other place.

Reply:
hmm I didn't understand. How "string" is related to `{category: category}`? And I don't understand how `{category: category}` wouldn't work. Can you send a small typescript playground example?

Reply:
Sure: https://www.typescriptlang.org/play/?ssl=28&ssc=1&pln=29&pc=1#code/PTAEGUEsFsAcBtIDNIFMAmoBGBPUAnVaAewDdIA7Ac1ADFL0B5WVfAQwBdj8BYAKH4ccLOgwDCxCukgdIkgM4AeACoA+UAF5QAb36h9oANoAFUJVABrVDmJJQygLoB+AFyipEqTLkUlykw6qANz8AL4hApGoAB6w3BygQiL0HpLSsgoAcqgA7irqWsqgAD46egYmZhSW1rb2zm7+xg6gMRyoUvKgxvjE0JDyqIqUSKygAKqq5fpO7uie6T7y2XmT06BuKfNp3lm5KgHBYfwnfADG8GzyXWKcqFTceLp8BmbobhQArtBYY1oAjBFXrA2IQKBw3D0+gMhrd2g98HhSp8pKgUBQMAVur1+oMAHSEeTEeCkVAAChR6DRlAwAEp1vIOJxIGdQOj0GTSGx4J9UJtxDsMr5FHD7o9VLSyi8DKEGUzZKz2StOdzefzUl4hct9qKETgJTpQLK+Mb+GcFAkQWCEloMTlQLrHmTaRF+I7EXj2WTtKArR0EqF6XwQPYABaoAioK6SRLh0CIDGgNhYMgR9DEVDyCgAcgS5rgkHgEYG2FQZzYn0G-BD7rwxCwACsywkSxxCHdMIgrL7LuZ602zi3fJAqXMFrtfKBq2A2FJQFTzZ8EBgqlwqtJyOhPtzfb0WPhZJnQDkZKHQAADUzmKw2OzKc-xyDdokuN13PWehjev3gtzaEduD+HB4iOAA0vqgv6gGQeCeJAaEganLWn5SMq2jwbSQA

Reply:
Can you add that to the rest just in case - or ensure its in the right place elsewhere in the codebase.